main.cpp

```diff
@@ -6,134 +6,714 @@
 {
-    double add(double lhs, double rhs) { return lhs + rhs; }
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value += static_cast<float>(*rhs.value);
+    return *this;
+}
 
-    double subtract(double lhs, double rhs) { return lhs - rhs; }
+FloatType& FloatType::add(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value += static_cast<float>(*rhs.value);
+    return *this;
+}
 
-    double multiply(double lhs, double rhs) { return lhs * rhs; }
+FloatType& FloatType::subtract(const float& rhs)
+{
+    *value -= rhs;
+    return *this;
+}
 
-    double divide(double lhs, double rhs)
+FloatType& FloatType::subtract(const FloatType& rhs)
+{
+    if (rhs.value == nullptr)
     {
-        if (rhs == 0)
-        {
-            std::cout << "divide by zero\n";
-        }
-        return lhs / rhs;
+        std::cout << "value null ptr\n";
+        return *this;
     }
-};
+    *value -= *rhs.value;
+    return *this;
+}
 
-struct IntType 
+FloatType& FloatType::subtract(const DoubleType& rhs)
 {
-    int add(int lhs, int rhs) { return lhs + rhs; }
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value -= static_cast<float>(*rhs.value);
+    return *this;
+}
 
-    int subtract(int lhs, int rhs) { return lhs - rhs; }
+FloatType& FloatType::subtract(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value -= static_cast<float>(*rhs.value);
+    return *this;
+}
 
-    int multiply(int lhs, int rhs) { return lhs * rhs; }
+FloatType& FloatType::multiply(const float& rhs)
+{
+    *value *= rhs;
+    return *this;
+}
 
-    int divide(int lhs, int rhs)
+FloatType& FloatType::multiply(const FloatType& rhs)
+{
+    if (rhs.value == nullptr)
     {
-        if (rhs == 0)
-        {
-            std::cerr << "divide by zero error\n";
-            return std::numeric_limits<int>::max(); // INFO we could debate for hours what is the proper return value for this case.  
-        }
-        return lhs / rhs;
+        std::cout << "value null ptr\n";
+        return *this;
     }
-};
+    *value *= *rhs.value;
+    return *this;
+}
 
-int main()
+FloatType& FloatType::multiply(const DoubleType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= static_cast<float>(*rhs.value);
+    return *this;
+}
+
+FloatType& FloatType::multiply(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= static_cast<float>(*rhs.value);
+    return *this;
+}
+
+FloatType& FloatType::divide(const float& rhs)
+{
+    if (rhs == 0.f)
+    {
+        std::cout << "divide by zero\n";
+        return *this;
```

Comment:
- [ ] floating point division by 0 is not an error.  you don't need to return yet.   Just warn the user.